src/engine/renderer/tr_light.cpp

```diff
@@ -743,122 +743,190 @@ void R_SetupLightProjection( trRefLight_t *light )
 
 		case refLightType_t::RL_PROJ:
 			{
-				int    i;
-				float  *proj = light->projectionMatrix;
-				vec3_t right, up, normal;
-				vec3_t start, stop;
-				vec3_t falloff;
-				float  falloffLen;
-				float  rLen;
-				float  uLen;
-				float  a, b, ofs, dist;
+				/* This was previously set as a vec4_t with lightProject[ x ] being 0,
+				but then the fourth component never contributed to any computation
+				and was always zeroed again after computation as a multiple of 0,
+				only the normal vec3 is used for computation. */
+				plane_t lightProject[ 4 ];
 
-				rLen = VectorNormalize2( light->l.projRight, right );
-				uLen = VectorNormalize2( light->l.projUp, up );
+				{
+					vec3_t right, up, normal;
 
-				CrossProduct( up, right, normal );
-				VectorNormalize( normal );
+					float rLen = VectorNormalize2( light->l.projRight, right );
+					float uLen = VectorNormalize2( light->l.projUp, up );
 
-				dist = DotProduct( light->l.projTarget, normal );
+					CrossProduct( up, right, normal );
+					VectorNormalize( normal );
 
-				if ( dist < 0 )
-				{
-					dist = -dist;
-					VectorInverse( normal );
-				}
+					vec_t dist = DotProduct( light->l.projTarget, normal );
+
+					if ( dist < 0 )
+					{
+						dist = -dist;
+						VectorInverse( normal );
+					}
 
-				VectorScale( right, ( 0.5f * dist ) / rLen, right );
-				VectorScale( up, - ( 0.5f * dist ) / uLen, up );
+					VectorScale( right, ( 0.5f * dist ) / rLen, right );
+					VectorScale( up, - ( 0.5f * dist ) / uLen, up );
 
-				vec4_t lightProject[ 4 ];
-				Vector4Set( lightProject[ 0 ], right[ 0 ], right[ 1 ], right[ 2 ], 0 );
-				Vector4Set( lightProject[ 1 ], up[ 0 ], up[ 1 ], up[ 2 ], 0 );
-				Vector4Set( lightProject[ 2 ], normal[ 0 ], normal[ 1 ], normal[ 2 ], 0 );
+					PlaneSet( lightProject[ 0 ], right[ 0 ], right[ 1 ], right[ 2 ], 0 );
+					PlaneSet( lightProject[ 1 ], up[ 0 ], up[ 1 ], up[ 2 ], 0 );
+					PlaneSet( lightProject[ 2 ], normal[ 0 ], normal[ 1 ], normal[ 2 ], 0 );
+				}
 
 				// now offset to center
-				vec4_t targetGlobal;
-				Vector4Set( targetGlobal,
-					light->l.projTarget[ 0 ], light->l.projTarget[ 1 ],
-					light->l.projTarget[ 2 ], 1 );
 				{
-					a = DotProduct4( targetGlobal, lightProject[ 0 ] );
-					b = DotProduct4( targetGlobal, lightProject[ 2 ] );
-					ofs = 0.5 - a / b;
+					/* This was previously set as a vec4_t with targetGlobal[ 3 ] being 1
+					but then the fourth component never contributed to any computation
+					and was always zeroed after computation as a multiple of 0. */
+					vec3_t targetGlobal;
+					VectorSet( targetGlobal,
+						light->l.projTarget[ 0 ], light->l.projTarget[ 1 ],
+						light->l.projTarget[ 2 ] );
+
+					{
+						/* This was previously doing a DotProduct4 on a vec4_t
```

Comment:
If you feel unsure about this refactoring, it would be better to avoid merging rather than write a bunch of comments apologizing for it. The comments don't help understand the code and just waste the reader's time.

Reply:
I'm fine with this refactoring as a separate commit.

My comments are not there to apologize for this, it's just that I don't know what was the code doing, and some stuff didn't make sense. If it was right it was mostly by luck, if it was wrong, this should not make it worse. I just believe the comments may be useful to someone else dealing with this code in the future.

I haven't spotted any regression.

Reply:
The point of a refactoring is to make the code nicer to work with before. The lengthy, irrelevant comments make it more difficult to read than it was before, so I will not LGTM it with those.

Reply:
Well, those comments are useful for me as they are kind of extended line-per-line commit messages, and I want to have those comments in the commit introducing the change if I have to bisect the code later. But I'm fine at removing them right after that. I added a commit to remove those comments.

Also, at the time I wrote this code, the legacy renderer was not re-enabled again. Now that it works again, I could test those changes with the dynamic light renderer and spotted no regression.